Build is failed on alpine linux, probably musl compatibility.

Hello, I'm trying to build on alpine linux. `cmake ..` didn't produce miracle-wm binary. I tried using these commands instead : 

```
$ cmake -B build
$ cmake --build build
```

Then I got these error messages : 

```
  8%] Building CXX object CMakeFiles/miracle-wm.dir/src/window_tree.cpp.o
/home/lidg/cmp/miracle-wm/src/window_tree.cpp: In member function 'void miracle::WindowTree::_handle_resize_request(const std::shared_ptr<miracle::Node>&, miracle::Direction, int)':
/home/lidg/cmp/miracle-wm/src/window_tree.cpp:548:38: error: 'floorf32' was not declared in this scope; did you mean 'floorf'?
  548 |         int height_for_others = (int)floorf32(-(float)resize_amount / static_cast<float>(nodes.size() - 1));
      |                                      ^~~~~~~~
      |                                      floorf
gmake[2]: *** [CMakeFiles/miracle-wm.dir/build.make:104: CMakeFiles/miracle-wm.dir/src/window_tree.cpp.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:100: CMakeFiles/miracle-wm.dir/all] Error 2
gmake: *** [Makefile:136: all] Error 2
```

Any clue to fix it ?

Also I need to comment `main.cpp` at line 60, if I don't I will get this error : 

```
  8%] Building CXX object CMakeFiles/miracle-wm.dir/src/main.cpp.o
/home/lidg/cmp/miracle-wm/src/main.cpp: In function 'int main(int, const char**)':
/home/lidg/cmp/miracle-wm/src/main.cpp:60:26: error: 'class miral::X11Support' has no member named 'default_to_enabled'
   60 |             X11Support{}.default_to_enabled(),
      |                          ^~~~~~~~~~~~~~~~~~
/home/lidg/cmp/miracle-wm/src/main.cpp:48:27: error: cannot convert '<brace-enclosed initializer list>' to 'std::initializer_list<std::function<void(mir::Server&)> >'
   48 |     return runner.run_with(
      |            ~~~~~~~~~~~~~~~^
   49 |         {
      |         ~                  
   50 |             window_managers,
      |             ~~~~~~~~~~~~~~~~
   51 |             WaylandExtensions{}
      |             ~~~~~~~~~~~~~~~~~~~
   52 |                 .enable(miral::WaylandExtensions::zwlr_layer_shell_v1)
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   53 |                 .enable(miral::WaylandExtensions::zwlr_foreign_toplevel_manager_v1)
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   54 |                 .enable(miral::WaylandExtensions::zxdg_output_manager_v1)
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   55 |                 .enable(miral::WaylandExtensions::zwp_virtual_keyboard_manager_v1)
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   56 |                 .enable(miral::WaylandExtensions::zwlr_virtual_pointer_manager_v1)
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   57 |                 .enable(miral::WaylandExtensions::zwp_input_method_manager_v2)
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   58 |                 .enable(miral::WaylandExtensions::zwlr_screencopy_manager_v1)
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   59 |                 .enable(miral::WaylandExtensions::ext_session_lock_manager_v1),
      |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   60 |             X11Support{}.default_to_enabled(),
      |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   61 |             config_keymap,
      |             ~~~~~~~~~~~~~~ 
   62 |             external_client_launcher,
      |             ~~~~~~~~~~~~~~~~~~~~~~~~~
   63 |             internal_client_launcher,
      |             ~~~~~~~~~~~~~~~~~~~~~~~~~
   64 |             display_configuration_options,
      |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   65 |             AddInitCallback(run_startup_apps)
      |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   66 |         });
      |         ~~                 
In file included from /home/lidg/cmp/miracle-wm/src/auto_restarting_launcher.h:6,
                 from /home/lidg/cmp/miracle-wm/src/main.cpp:5:
/usr/include/miral/miral/runner.h:78:78: note:   initializing argument 1 of 'int miral::MirRunner::run_with(std::initializer_list<std::function<void(mir::Server&)> >)'
   78 | initializer_list<std::function<void(::mir::Server&)>> options) -> int;
      | ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~

gmake[2]: *** [CMakeFiles/miracle-wm.dir/build.make:76: CMakeFiles/miracle-wm.dir/src/main.cpp.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:100: CMakeFiles/miracle-wm.dir/all] Error 2
gmake: *** [Makefile:136: all] Error 2
```

--- a/src/node.cpp
+++ b/src/node.cpp
@@ -90,7 +90,7 @@ InsertNodeInternalResult insert_node_internal(
     std::function<int(int)> const& get_node_position,
     std::function<void(int, int, int)> const& set_node_size_position)
 {
-    int new_item_size = floorf64((double)lane_size / (double)(node_count + 1));
+    int new_item_size = floor((double)lane_size / (double)(node_count + 1));
     int new_item_position = lane_pos + index * new_item_size;
 
     int size_lost = 0;
@@ -105,7 +105,7 @@ InsertNodeInternalResult insert_node_internal(
         // (meaning that larger nodes give more width, and lesser nodes give less width)
         double percent_size_lost =
             ((double)node_size / (double)lane_size);
-        int width_to_lose = (int)floorf64(percent_size_lost * new_item_size);
+        int width_to_lose = (int)floor(percent_size_lost * new_item_size);
         size_lost += width_to_lose;
 
         if (i == index)
@@ -314,14 +314,16 @@ void Node::set_logical_area(geom::Rectangle const& target_rect)
         // However, the "non-main-axis" dimension will be consistent across each node.
         auto placement_area = get_logical_area();
         auto target_placement_area = get_logical_area_internal(target_rect);
+        std::vector<geom::Rectangle> pending_size_updates;
         if (direction == NodeLayoutDirection::horizontal)
         {
+            int total_width = 0;
             for (size_t idx = 0; idx < sub_nodes.size(); idx++)
             {
                 auto item = sub_nodes[idx];
                 auto item_rect = item->get_logical_area();
-                float percent_width_taken = (float)item_rect.size.width.as_int() / (float)placement_area.size.width.as_int();
-                int new_width = (int)ceilf((float)target_placement_area.size.width.as_int() * percent_width_taken);
+                double percent_width_taken = (double)item_rect.size.width.as_int() / (double)placement_area.size.width.as_int();
+                int new_width = (int)ceil((double)target_placement_area.size.width.as_int() * percent_width_taken);
 
                 geom::Rectangle new_item_rect;
                 new_item_rect.size = geom::Size{
@@ -337,24 +339,32 @@ void Node::set_logical_area(geom::Rectangle const& target_rect)
                 }
                 else
                 {
-                    auto prev_rect = sub_nodes[idx - 1]->get_logical_area();
+                    auto const& prev_rect = pending_size_updates[idx - 1];
                     new_item_rect.top_left = geom::Point{
                         geom::X{prev_rect.top_left.x.as_int() + prev_rect.size.width.as_int()},
                         target_placement_area.top_left.y
                     };
                 }
 
-                item->set_logical_area(new_item_rect);
+                pending_size_updates.push_back(new_item_rect);
+                total_width += new_width;
+            }
+
+            if (!pending_size_updates.empty())
+            {
+                int leftover_width = target_placement_area.size.width.as_int() - total_width;
+                pending_size_updates.back().size.width = geom::Width {pending_size_updates.back().size.width.as_int() + leftover_width};
             }
         }
         else
         {
+            int total_height = 0;
             for (size_t idx = 0; idx < sub_nodes.size(); idx++)
             {
                 auto item = sub_nodes[idx];
                 auto item_rect = item->get_logical_area();
-                float percent_height_taken = static_cast<float>(item_rect.size.height.as_int()) / placement_area.size.height.as_int();
-                int new_height = (int)floorf((float)target_placement_area.size.height.as_int() * percent_height_taken);
+                double percent_height_taken = static_cast<double>(item_rect.size.height.as_int()) / placement_area.size.height.as_int();
+                int new_height = (int)floor((double)target_placement_area.size.height.as_int() * percent_height_taken);
 
                 geom::Rectangle new_item_rect;
                 new_item_rect.size = geom::Size{
@@ -370,17 +380,30 @@ void Node::set_logical_area(geom::Rectangle const& target_rect)
                 }
                 else
                 {
-                    auto prev_rect = sub_nodes[idx - 1]->get_logical_area();
+                    auto const& prev_rect = pending_size_updates[idx - 1];
                     new_item_rect.top_left = geom::Point{
                         target_placement_area.top_left.x,
                         geom::Y{prev_rect.top_left.y.as_int() + prev_rect.size.height.as_int()},
                     };
                 }
 
-                item->set_logical_area(new_item_rect);
+                pending_size_updates.push_back(new_item_rect);
+                total_height += new_height;
+            }
+
+            if (!pending_size_updates.empty())
+            {
+                int leftover_height = target_placement_area.size.height.as_int() - total_height;
+                pending_size_updates.back().size.height = geom::Height {pending_size_updates.back().size.height.as_int() + leftover_height};
             }
         }
+
+        for (size_t i = 0; i < sub_nodes.size(); i++)
+        {
+            sub_nodes[i]->set_logical_area(pending_size_updates[i]);
+        }
     }
+
 
     // Important that we update the area _after_ changes have taken place!
     logical_area = target_rect;
@@ -540,8 +563,8 @@ geom::Rectangle Node::_get_visible_from_logical(
     bool has_bottom_neighbor,
     std::shared_ptr<MiracleConfig> const& config)
 {
-    int half_gap_x = has_right_neighbor ? (int)(ceilf((float) config->get_inner_gaps_x() / 2.f)) : 0;
-    int half_gap_y = has_bottom_neighbor ? (int)(ceilf((float) config->get_inner_gaps_y() / 2.f)) : 0;
+    int half_gap_x = has_right_neighbor ? (int)(ceil((double) config->get_inner_gaps_x() / 2.0)) : 0;
+    int half_gap_y = has_bottom_neighbor ? (int)(ceil((double) config->get_inner_gaps_y() / 2.0)) : 0;
     return {
         geom::Point{
             logical_area.top_left.x.as_int(),

--- a/src/tree.cpp
+++ b/src/tree.cpp
@@ -541,7 +541,8 @@ void Tree::_handle_resize_request(
     std::vector<geom::Rectangle> pending_node_resizes;
     if (is_vertical)
     {
-        int height_for_others = (int)floorf32(-(float)resize_amount / static_cast<float>(nodes.size() - 1));
+        int height_for_others = (int)floor(-(double)resize_amount / static_cast<double>(nodes.size() - 1));
+        int total_height = 0;
         for (size_t i = 0; i < nodes.size(); i++)
         {
             auto other_node = nodes[i];
@@ -563,12 +564,18 @@ void Tree::_handle_resize_request(
                 return;
             }
 
+            total_height += other_rect.size.height.as_int();
             pending_node_resizes.push_back(other_rect);
         }
+
+        // Due to some rounding errors, we may have to extend the final node
+        int leftover_height = parent->get_logical_area().size.height.as_int() - total_height;
+        pending_node_resizes.back().size.height = geom::Height{pending_node_resizes.back().size.height.as_int() + leftover_height};
     }
     else
     {
-        int width_for_others = (int)floorf((float)-resize_amount / static_cast<float>(nodes.size() - 1));
+        int width_for_others = (int)floor((double)-resize_amount / static_cast<double>(nodes.size() - 1));
+        int total_width = 0;
         for (size_t i = 0; i < nodes.size(); i++)
         {
             auto other_node = nodes[i];
@@ -590,8 +597,13 @@ void Tree::_handle_resize_request(
                 return;
             }
 
+            total_width += other_rect.size.width.as_int();
             pending_node_resizes.push_back(other_rect);
         }
+
+        // Due to some rounding errors, we may have to extend the final node
+        int leftover_width = parent->get_logical_area().size.width.as_int() - total_width;
+        pending_node_resizes.back().size.width = geom::Width {pending_node_resizes.back().size.width.as_int() + leftover_width};
     }
 
     for (size_t i = 0; i < nodes.size(); i++)

